compiler/utils/bitsets.nim

```diff
@@ -11,62 +11,116 @@
                               # cross-compiling; uint would be more efficient
                               # however
+  
+  # XXX: aliasing `openArray` this way may currently trip up the compiler, so 
+  # use `openArray` directly for now
+  #openArray[ElemType]* = openArray[ElemType] ## A view into a bit set. This doesn't
+                                     ## necessarily have to be a `TBitSet`,
+                                     ## but can be any byte sequence storing
+                                     ## representing a bit set
+
 const
   ElemSize* = 8
   One = ElemType(1)
   Zero = ElemType(0)
+  AllOne = ElemType(0xFF)
 
 template modElemSize(arg: untyped): untyped = arg and 7
 template divElemSize(arg: untyped): untyped = arg shr 3
 
-proc bitSetIn*(x: TBitSet, e: BiggestInt): bool =
+func bitSetIn*(x: openArray[ElemType], e: BiggestInt): bool =
+  ## Tests if `e` is an element of `x`
   result = (x[int(e.divElemSize)] and (One shl e.modElemSize)) != Zero
 
-proc bitSetIncl*(x: var TBitSet, elem: BiggestInt) =
+func bitSetIncl*(x: var openArray[ElemType], elem: BiggestInt) =
+  ## If `elem` is not in set `x` yet, includes the element
   assert(elem >= 0)
   x[int(elem.divElemSize)] = x[int(elem.divElemSize)] or
       (One shl elem.modElemSize)
 
-proc bitSetExcl*(x: var TBitSet, elem: BiggestInt) =
+func bitSetExcl*(x: var openArray[ElemType], elem: BiggestInt) =
+  ## If `elem` is in set `x`, removes it, otherwise does nothing
   x[int(elem.divElemSize)] = x[int(elem.divElemSize)] and
       not(One shl elem.modElemSize)
 
-proc bitSetInit*(b: var TBitSet, length: int) =
+func bitSetInclRange*(x: var openArray[ElemType], s: Slice[BiggestInt]) =
+  ## Includes elements in the slice `r` that aren't part of `x` yet into `x`
+  if unlikely(s.a > s.b): return # Do nothing for empty slices
+
+  # This functions aims to be more efficient than a for loop with `bitSetIncl`
+
+  let start = int(s.a.divElemSize)
+  let last = int(s.b.divElemSize)
+
+  let firstBit = s.a.modElemSize
+  let endBit = s.b.modElemSize + 1 # lastBit + 1
+
+  # The position of the last bit + 1 in the start element
+  let startElemEndBit =
+    if start < last: ElemSize
+    else: int(endBit)
+
+  # Calculate the intersection between the bit ranges `0..lastBit` and `firstBit..high`
+  let startElemBits = ((One shl startElemEndBit) - 1) and (AllOne shl firstBit)
+
+  x[start] = x[start] or startElemBits
+
+  for i in (start+1)..<last:
+    x[i] = AllOne
+
+  if start < last:
+    x[last] = x[last] or ((One shl endBit) - 1)
+
+func bitSetInit*(b: var TBitSet, length: int) =
+  ## Creates a new bitset in `b` with a size-in-bytes of `length * ElemSize` .
+  ## The resulting bitset allows for elements in the
+  ## range `0..<(length * ElemSize * 8)`
   newSeq(b, length)
 
-proc bitSetUnion*(x: var TBitSet, y: TBitSet) =
+func bitSetUnion*(x: var openArray[ElemType], y: openArray[ElemType]) =
+  ## Calculate the union between sets `x` and `y
```

Comment:
```suggestion
  ## Calculate the union between sets `x` and `y`
```

That should fix docgen...

Reply:
Ah, thanks. I wasn't able to figure out why docgen was failing while all other test passed.

Should I revert the `TBitSetView` -> `openArray[ElemType]` change then, as it didn't cause the issue?

Reply:
Feel free to